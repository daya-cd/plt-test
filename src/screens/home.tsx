import React, {FC, useCallback, useEffect, useState} from 'react';
import {SafeAreaView, StyleSheet, View, Text, FlatList} from 'react-native';
import {Subheading} from 'react-native-paper';
import Products from '../components/products';
import {IProduct} from '../shared/interfaces/product';

const Home: FC = () => {
  const [products, setProducts] = useState<IProduct[]>([]);
  const [error, setError] = useState<string | null>();

  const fetcData = async () => {
    const resp = await fetch(
      'https://my-json-server.typicode.com/benirvingplt/products/products',
    );
    if (!resp.ok) {
      setError('Something went wrong');
    } else {
      const data = await resp.json();
      setProducts(data);
      setError(null);
    }
  };

  useEffect(() => {
    fetcData();
  }, []);

  const renderItem = useCallback(({item}) => <Products product={item} />, []);
  const keyExtractor = useCallback(item => item.id.toString(), []);

  return (
    <SafeAreaView style={styles.container}>
      <View>
        {!products && <Text> Loading...</Text>}
        {products && (
          <FlatList
            data={products}
            keyExtractor={keyExtractor}
            renderItem={renderItem}
          />
        )}
        {error && (
          <View style={styles.noProduct}>
            <Subheading>Something went wrong in API call</Subheading>
          </View>
        )}
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  noProduct: {
    margin: 20,
  },
});

export default Home;
