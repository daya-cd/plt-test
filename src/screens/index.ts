import Home from './home';
import Basket from './basket';
import ProductDetail from './productDetail';

export {Home, Basket, ProductDetail};
