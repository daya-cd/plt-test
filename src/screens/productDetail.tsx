import React from 'react';
import {View, SafeAreaView, StyleSheet, Image} from 'react-native';
import {
  Card,
  Title,
  Button,
  Portal,
  Dialog,
  Paragraph,
} from 'react-native-paper';
import {useDispatch} from 'react-redux';
import {bindActionCreators} from 'redux';
import {actionCreators} from '../shared/store';
import {IProduct} from '../shared/interfaces/product';

const ProductDetail = ({route}) => {
  const dispatch = useDispatch();
  const {addProduct} = bindActionCreators(actionCreators, dispatch);
  const [visible, setVisible] = React.useState(false);

  const hideDialog = () => setVisible(false);

  const addToCart = (product: IProduct): void => {
    addProduct(product);
    setVisible(true);
  };
  const {productDetail} = route.params;
  return (
    <SafeAreaView style={styles.container}>
      <View>
        <Image
          source={{uri: productDetail.img}}
          style={styles.imageContainer}
          resizeMode="center"
        />
        <Card.Title
          titleNumberOfLines={2}
          titleStyle={styles.title}
          title={productDetail.name}
          subtitle={'Colour: ' + productDetail.colour}
          right={() => <Title>Â£{productDetail.price}</Title>}
          rightStyle={styles.rightStyle}
        />
        <Button
          mode="contained"
          testID="addCartButton"
          onPress={() => addToCart(productDetail)}>
          ADD TO CART
        </Button>
        <Portal>
          <Dialog visible={visible} onDismiss={hideDialog}>
            <Dialog.Title>Update:</Dialog.Title>
            <Dialog.Content>
              <Paragraph>Item has been added to your basket.</Paragraph>
            </Dialog.Content>
            <Dialog.Actions>
              <Button onPress={hideDialog}>Done</Button>
            </Dialog.Actions>
          </Dialog>
        </Portal>
      </View>
    </SafeAreaView>
  );
};

export default ProductDetail;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    margin: 10,
  },
  imageContainer: {
    height: 400,
  },
  title: {
    fontSize: 14,
  },
  rightStyle: {
    paddingRight: 20,
  },
});
