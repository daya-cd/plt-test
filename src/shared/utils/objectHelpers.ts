const isEmpty = (obj: object) => {
  return Object.keys(obj).length === 0;
};
export {isEmpty};
