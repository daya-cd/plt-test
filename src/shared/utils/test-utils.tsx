import React, {ReactElement} from 'react';
import {
  render as rtlRender,
  RenderOptions,
} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {Provider as PaperProvider} from 'react-native-paper';
import {store} from '../store';

export interface WrapperProps {
  children: ReactElement;
}

const render = (ui: ReactElement, renderOptions?: RenderOptions) => {
  const Wrapper = ({children}: WrapperProps): ReactElement => {
    return (
      <Provider store={store}>
        <PaperProvider>{children}</PaperProvider>
      </Provider>
    );
  };
  return rtlRender(ui, {wrapper: Wrapper, ...renderOptions});
};

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {render};
