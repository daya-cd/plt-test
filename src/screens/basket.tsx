/* eslint-disable no-shadow */
import React from 'react';
import {
  SafeAreaView,
  StyleSheet,
  Image,
  ScrollView,
  TouchableOpacity,
  View,
} from 'react-native';
import {Colors, List, Subheading} from 'react-native-paper';
import {useSelector, useDispatch} from 'react-redux';
import {RootState} from '../shared/store/reducers';
import {bindActionCreators} from 'redux';
import {actionCreators} from '../shared/store';
import {IProduct} from '../shared/interfaces/product';
import {isEmpty} from '../shared/utils/objectHelpers';

const Basket = () => {
  const state = useSelector((state: RootState) => state.shop);
  const dispatch = useDispatch();
  const {removeProduct} = bindActionCreators(actionCreators, dispatch);
  const removeItemFromCart = (product: IProduct): void => {
    removeProduct(product);
  };
  if (isEmpty(state.items)) {
    return (
      <View style={styles.noProduct}>
        <Subheading>
          No products in the basket. Please select from home
        </Subheading>
      </View>
    );
  }
  return (
    <SafeAreaView style={styles.container}>
      <ScrollView>
        {state &&
          state.items.map((items, i) => (
            <List.Item
              key={i}
              title={items.name}
              titleNumberOfLines={2}
              description={
                'Price: Â£' + items.price + '  ' + 'Colour: ' + items.colour
              }
              left={() => (
                <Image
                  source={{uri: items.img}}
                  style={styles.imageContainer}
                  resizeMode="contain"
                />
              )}
              right={props => (
                <TouchableOpacity
                  onPress={() => removeItemFromCart(items)}
                  testID="deleteIcon">
                  <List.Icon {...props} color={Colors.redA700} icon="delete" />
                </TouchableOpacity>
              )}
            />
          ))}
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  imageContainer: {
    width: 80,
    height: 80,
  },
  noProduct: {
    margin: 20,
  },
});

export default Basket;
