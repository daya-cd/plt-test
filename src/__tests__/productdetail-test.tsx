import 'react-native';
import React from 'react';
import * as redux from 'react-redux';
import {ProductDetail} from '../screens';
import {fireEvent, render} from '../shared/utils/test-utils';

const defaultProps = {
  params: {
    productDetail: {
      id: 2,
      colour: 'Stone',
      name: 'Stone Ribbed Strappy Cut Out Detail Bodycon Dress',
      price: 4,
      img: 'https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024',
    },
  },
};
const setup = () => {
  return render(<ProductDetail route={defaultProps} />);
};

describe('renders product detail view', () => {
  const useDispatchMock = jest.spyOn(redux, 'useDispatch');
  beforeEach(() => {
    useDispatchMock.mockClear();
  });
  it('renders correctly', () => {
    const tree = setup().toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders product detail name correctly', async () => {
    const {findByText} = setup();
    const productName = await findByText(
      /Stone Ribbed Strappy Cut Out Detail Bodycon Dress/i,
    );
    expect(productName).toBeTruthy();
  });

  it('fires addProduct dispatch action', () => {
    const addProduct = jest.fn();
    useDispatchMock.mockReturnValue(addProduct);
    const {getByTestId} = setup();
    expect(addProduct).not.toHaveBeenCalled();
    fireEvent.press(getByTestId('addCartButton'));
    expect(addProduct).toHaveBeenCalled();
  });
});
