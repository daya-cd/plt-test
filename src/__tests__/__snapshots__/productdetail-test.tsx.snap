// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders product detail view renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
        "margin": 10,
      }
    }
  >
    <View>
      <Image
        resizeMode="center"
        source={
          Object {
            "uri": "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024",
          }
        }
        style={
          Object {
            "height": 400,
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingLeft": 16,
            },
            Object {
              "minHeight": 72,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            numberOfLines={2}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    Array [
                      Object {
                        "minHeight": 30,
                      },
                      Object {
                        "marginBottom": 0,
                      },
                      Object {
                        "fontSize": 14,
                      },
                    ],
                  ],
                ],
              ]
            }
          >
            Stone Ribbed Strappy Cut Out Detail Bodycon Dress
          </Text>
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.54)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 12,
                      "letterSpacing": 0.4,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Array [
                      Object {
                        "marginVertical": 0,
                        "minHeight": 20,
                      },
                      undefined,
                    ],
                  ],
                ],
              ]
            }
          >
            Colour: Stone
          </Text>
        </View>
        <View
          style={
            Object {
              "paddingRight": 20,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            Â£
            4
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#6200ee",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 2,
            "minWidth": 64,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.75,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 1.5,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
          testID="addCartButton"
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    undefined,
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#ffffff",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              ADD TO CART
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
